### ISSUE TITLE
typescript information

### ISSUE THREAD

**Issue by ofbeaton** [leny/react-use-storage#1]
**tl;dr: there's no type info for Typescript. I include one that users of this project can use in their projects at the bottom of my post in case an official one is never published.**

Hi, I'm just a beginner learning React + Typescript, so please have a bit of patience with me.

Thank you for the nice module.

I had trouble using your module in typescript because it lacks type information. I spent all day trying to figure out how to resolve this, sadly the docs on how to fix this are not great.

I gathered that it is missing type information in the module itself or in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) (NPM @types).

### Publishing Type information

According to Typescript's Type [Publishing](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) page
```
There are two main ways you can publish your declaration files to npm:

    1. bundling with your npm package, or
    2. publishing to the @types organization on npm.
If your package is written in TypeScript then the first approach is favored. Use the --declaration flag to generate declaration files. This way, your declarations and JavaScript will always be in sync.

If your package is not written in TypeScript then the second is the preferred approach.
```

Since this is a javascript module, I'm going to assume you have no plans to rewrite it in Typescript, so the DefinitelyTyped avenue is the recommended one. 

Furthermore I noticed that DefinitelyTyped states that type PRs coming from module authors are prioritized.

Looking at the process for writing a type module for inclusion in @types looks super scary to me as a beginner and there is a ton there I can't even begin to understand.

### Fixing the types by hand in an application

I found an article that helped me [Migrating to Typescript: Write a declaration file for a third-party NPM module](https://medium.com/@chris_72272/migrating-to-typescript-write-a-declaration-file-for-a-third-party-npm-module-b1f75808ed2) write type information,

In my create-react-app project I made a file

_src/types/react-use-storage/index.d.ts_
```typescript
declare module "react-use-storage" {
  export function useLocalStorage<T>(
    key: string,
    defaultValue: T
  ): [T, (updatedValue: T) => void];
  export function useSessionStorage<T>(
    key: string,
    defaultValue: T
  ): [T, (updatedValue: T) => void];
}
```

I did not need to edit my _tsconfig.json_ as it already contains the line `"include": ["src"]`

This will fix it in a local project. Looking at the official docs I can tell this is not how they want you to do things (with `declare module`). But all my efforts to make one the official way failed.

### Moving forwards

If I still use _react-use-storage_ in a several months when I feel more comfortable in javascript and typescript, and if no type information exists at that time, I may revisit this and try to submit a type for this module to DefinitelyTyped, but right now that is far too ambitious for me.

**Comment by leny** [comment]
Hi Finlay,

Thanks for this well-detailed issue!

I’m not (yet?) a big user of typescript, so I always forgot the address the types export when creating a node package. Shame on me.

I will definitely add types to this package ASAP, probably this week.

Have a nice Sunday!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by leny** [comment]
I got some time to dig around that.

I've just published version `0.3.1`, including typescript types declaration.

It's my first package including it, hope I don't broke anything ;)

**Comment by ofbeaton** [comment]
upgraded to `0.3.1`, works great. tsc compiled my code no problem without a local type def. 

Thank you so much for the quick turnaround, very much appreciated.
